### PR TITLE
Refactor dependencies for external SAT solvers

### PR DESCRIPTION
**Pull request by nafur**
This PR refactors how we obtain, build and use the external SAT solvers used by CVC4: CaDiCaL, CryptoMiniSat and Kissat.
All three contrib scripts are removed and instead an external project is integrated into the cmake find scripts.

### CODE REVIEW

File: cmake/FindCaDiCaL.cmake [hunk 1]
```diff
@@ -13,15 +13,80 @@
...
+    check_system_version("CaDiCaL")
+endif()
+
+if(NOT CaDiCaL_FOUND_SYSTEM)
+    include(CheckSymbolExists)
+    include(ExternalProject)
+
+    fail_if_include_missing("sys/resource.h" "CaDiCaL")
+
+    # TODO(mpreiner): Use a more recent version?
```
Comment (mpreiner) [inline cmake/FindCaDiCaL.cmake:45]: Yes, but we'll do that in a separate commit (we'll do that when preparing for SMT-COMP).

File: cmake/FindCaDiCaL.cmake [hunk 2]
```diff
@@ -13,15 +13,80 @@
...
+
+    ExternalProject_Add(
+        CaDiCaL-EP
+        PREFIX ${DEPS_PREFIX}
+        BUILD_IN_SOURCE ON
+        URL https://github.com/arminbiere/cadical/archive/refs/tags/rel-${CaDiCaL_VERSION}.tar.gz
+        URL_HASH SHA1=9de1176737b74440921ba86395fe5edbb3b131eb
+        CONFIGURE_COMMAND mkdir -p <SOURCE_DIR>/build
+        # avoid configure script, prepare the makefile manually
+        COMMAND sed -e "s,@CXX@,${CMAKE_CXX_COMPILER}," -e "s,@CXXFLAGS@,${CXXFLAGS}," -e "s,@MAKEFLAGS@,-j1," <SOURCE_DIR>/makefile.in > <SOURCE_DIR>/build/makefile
```
Comment (mpreiner) [inline cmake/FindCaDiCaL.cmake:66]: Is `-jN` passed to external projects when we build CVC4 with `make -jN`?

Comment (mpreiner) [inline cmake/FindCaDiCaL.cmake:66]: Indent `COMMAND` since it belongs to `CONFIGURE_COMMAND`?

File: cmake/FindCaDiCaL.cmake [hunk 3]
```diff
@@ -13,15 +13,80 @@
...
+        BUILD_COMMAND make -C <SOURCE_DIR>/build libcadical.a
+        INSTALL_COMMAND ${CMAKE_COMMAND} -E copy <SOURCE_DIR>/build/libcadical.a <INSTALL_DIR>/lib/libcadical.a
+        COMMAND ${CMAKE_COMMAND} -E copy <SOURCE_DIR>/src/cadical.hpp <INSTALL_DIR>/include/cadical.hpp
```
Comment (mpreiner) [inline cmake/FindCaDiCaL.cmake:69]: Ditto.

File: cmake/FindCryptoMiniSat.cmake [hunk 4]
```diff
@@ -13,16 +13,78 @@
 # CryptoMiniSat_INCLUDE_DIR - the CryptoMiniSat include directory
 # CryptoMiniSat_LIBRARIES - Libraries needed to use CryptoMiniSat
 
+include(deps-helper)
 
 find_path(CryptoMiniSat_INCLUDE_DIR NAMES cryptominisat5/cryptominisat.h)
```
Comment (mpreiner) [inline cmake/FindCryptoMiniSat.cmake:18]: CryptoMiniSat exports CMake targets when installed. We should use `find_package(cryptominisat5)`with the specific version we want in this case.

File: cmake/FindCryptoMiniSat.cmake [hunk 5]
```diff
@@ -13,16 +13,78 @@
...
+    )
+    ExternalProject_Add_Step(
+        CryptoMiniSat-EP
+        cleanup
+        DEPENDEES install
+        COMMAND ${CMAKE_COMMAND} -E remove <BINARY_DIR>/cryptominisat5_simple
+        COMMAND ${CMAKE_COMMAND} -E remove <INSTALL_DIR>/bin/cryptominisat5_simple
+    )
+
+    set(CryptoMiniSat_INCLUDE_DIR "${DEPS_BASE}/include/")
```
Comment (mpreiner) [inline cmake/FindCryptoMiniSat.cmake:70]: Again, we should use the installed CMake target for CryptoMiniSat.

**Review by mpreiner — CHANGES_REQUESTED** [review]
Make sure to format to 80 chars.

File: cmake/FindCaDiCaL.cmake (hunk 2 above)
Comment (nafur) [reply cmake/FindCaDiCaL.cmake:66]: A `COMMAND` generally belongs to the `X_COMMAND` that comes right before it. (see https://cmake.org/cmake/help/latest/module/ExternalProject.html#external-project-definition at "Miscellaneous Options:")

Comment (mpreiner) [reply cmake/FindCaDiCaL.cmake:66]: I know, that's why I suggested to indent it to make it clearer, but if you prefer it this way I'm fine with it.

Comment (nafur) [reply cmake/FindCaDiCaL.cmake]: If used properly, the external projects share the jobserver of the whole build. (Aka: yes).
In this particular case, I explicitly set `-j1` in the makefile, and also called `make` manually. I don't know why... it is fixed now.

File: cmake/FindCryptoMiniSat.cmake (hunk 4 above)
Comment (nafur) [reply cmake/FindCryptoMiniSat.cmake:18]: You are right. I'm working on it (see https://github.com/msoos/cryptominisat/pull/645)

File: cmake/FindCryptoMiniSat.cmake (hunk 5 above)
Comment (nafur) [reply cmake/FindCryptoMiniSat.cmake:70]: I'm not sure about this. The naive way would be to call `find_package` at this point in time. This does not work, though, as this searches for the export file that is only generated after CryptoMiniSat has been installed (i.e. at build time).

Comment (mpreiner) [reply cmake/FindCryptoMiniSat.cmake:67]: Ah yes. Hmm, well then we have to do it this way.

**Review by mpreiner — COMMENTED** [review]
Can you break lines to 80 chars and make consistent indentation (2 vs. 4 spaces)? I think until now we mostly used 2 spaces for indentation in CMake files.

**Comment by nafur** [comment]
> Can you break lines to 80 chars and make consistent indentation (2 vs. 4 spaces)? I think until now we mostly used 2 spaces for indentation in CMake files.

I've used `cmake-format --tab-size 2 --dangle-parens YES --first-comment-is-literal` now.

**Comment by mpreiner** [comment]
> > Can you break lines to 80 chars and make consistent indentation (2 vs. 4 spaces)? I think until now we mostly used 2 spaces for indentation in CMake files.
> 
> I've used `cmake-format --tab-size 2 --dangle-parens YES --first-comment-is-literal` now.

Nice, didn't know about cmake-format!

File: src/CMakeLists.txt [hunk 6]
```diff
@@ -1128,20 +1128,17 @@ if(USE_ABC)
   target_include_directories(cvc4 PRIVATE ${ABC_INCLUDE_DIR})
 endif()
 if(USE_CADICAL)
-  target_link_libraries(cvc4 ${CaDiCaL_LIBRARIES})
-  target_include_directories(cvc4 PRIVATE ${CaDiCaL_INCLUDE_DIR})
+  target_link_libraries(cvc4 CaDiCaL)
```
Comment (mpreiner) [inline src/CMakeLists.txt:1131]: Link PRIVATE?

File: src/CMakeLists.txt [hunk 7]
```diff
@@ -1128,20 +1128,17 @@ if(USE_ABC)
...
 endif()
 if(USE_CLN)
   target_link_libraries(cvc4 ${CLN_LIBRARIES})
   target_include_directories(cvc4 PUBLIC $<BUILD_INTERFACE:${CLN_INCLUDE_DIR}>)
 endif()
 if(USE_CRYPTOMINISAT)
-  target_link_libraries(cvc4 ${CryptoMiniSat_LIBRARIES})
-  target_include_directories(cvc4 PRIVATE ${CryptoMiniSat_INCLUDE_DIR})
+  target_link_libraries(cvc4 CryptoMiniSat)
```
Comment (mpreiner) [inline src/CMakeLists.txt:1138]: Ditto.

File: src/CMakeLists.txt [hunk 8]
```diff
@@ -1128,20 +1128,17 @@ if(USE_ABC)
...
 endif()
 if(USE_KISSAT)
-  target_link_libraries(cvc4 ${Kissat_LIBRARIES})
-  target_include_directories(cvc4 PRIVATE ${Kissat_INCLUDE_DIR})
+  target_link_libraries(cvc4 Kissat)
```
Comment (mpreiner) [inline src/CMakeLists.txt:1141]: Ditto.

File: src/CMakeLists.txt (hunk 6 above)
Comment (nafur) [reply src/CMakeLists.txt:1131]: This needs to be tone for all calls at once... see #6218.

File: cmake/FindCaDiCaL.cmake [hunk 9]
```diff
@@ -1,27 +1,99 @@
...
-## in the top-level source directory and their institutional affiliations.
-## All rights reserved.  See the file COPYING in the top-level source
-## directory for licensing information.
-##
+# FindCaDiCaL.cmake
+# Top contributors (to current version):
+#   Mathias Preiner
+# This file is part of the CVC4 project.
+# Copyright (c) 2009-2021 by the authors listed in the file AUTHORS
+# in the top-level source directory and their institutional affiliations.
```
Comment (mpreiner) [inline cmake/FindCaDiCaL.cmake:7]: Oh, this change will break the update copyright script as it expects `##`. Can you revert the change?

File: cmake/FindCryptoMiniSat.cmake [hunk 10]
```diff
@@ -1,28 +1,102 @@
...
-##
+# FindCryptoMiniSat.cmake
+# Top contributors (to current version):
+#   Mathias Preiner
+# This file is part of the CVC4 project.
+# Copyright (c) 2009-2021 by the authors listed in the file AUTHORS
+# in the top-level source directory and their institutional affiliations.
+# All rights reserved.  See the file COPYING in the top-level source
+# directory for licensing information.
+#
```
Comment (mpreiner) [inline cmake/FindCryptoMiniSat.cmake:10]: Ditto

File: cmake/FindKissat.cmake [hunk 11]
```diff
@@ -1,27 +1,82 @@
...
-##
+# FindKissat.cmake
+# Top contributors (to current version):
+#   Aina Niemetz
+# This file is part of the CVC4 project.
+# Copyright (c) 2009-2021 by the authors listed in the file AUTHORS
+# in the top-level source directory and their institutional affiliations.
+# All rights reserved.  See the file COPYING in the top-level source
+# directory for licensing information.
+#
```
Comment (mpreiner) [inline cmake/FindKissat.cmake:10]: Ditto

File: cmake/FindCaDiCaL.cmake (hunk 9 above)
Comment (nafur) [reply cmake/FindCaDiCaL.cmake]: done,

File: cmake/FindCaDiCaL.cmake [hunk 12]
```diff
@@ -1,4 +1,4 @@
-#####################
+######################
```
Comment (mpreiner) [inline cmake/FindCaDiCaL.cmake:1]:
```suggestion
#####################
```

File: cmake/FindCryptoMiniSat.cmake [hunk 13]
```diff
@@ -1,4 +1,4 @@
-#####################
+######################
```
Comment (mpreiner) [inline cmake/FindCryptoMiniSat.cmake:1]:
```suggestion
#####################
```

File: cmake/FindKissat.cmake [hunk 14]
```diff
@@ -1,4 +1,4 @@
-#####################
+######################
```
Comment (mpreiner) [inline cmake/FindKissat.cmake:1]:
```suggestion
#####################
```

**Review by mpreiner — APPROVED** [review]
